### PR TITLE
Assign machine- and RUN-specific resources

### LINKED ISSUE #177: Optimize resources for all resolutions on supported platforms
Test the following resolutions on WCOSS, Hera, and Orion to better define resource settings:

- C768
- C384
- C192
- C96
- C48

Special attention needs to be paid to lower resolutions for the fcst model; the atmospheric and wave parts are unbalanced.

### PR DESCRIPTION
**Pull request by DavidHuber-NOAA**
# Description
This redefines resource variables so they capture the `RUN` or `CDUMP` that they are valid for.  Additionally, machine-specific resources are moved out of `config.resources` and placed in respective `config.resources.{machine}` files.

Resolves #177 #2672
Also helps address 2092 for WCOSS2 and Gaea
# Type of change

- Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
Created multiple experiments on Hercules.  Additional tests to follow.
- [x] C48_S2SW test on Hercules
- [x] C48_S2SW test on WCOSS2
- [x] C96C48_hybatmDA on Hercules
- [x] C96C48_hybatmDA on Hera
- [x] C96C48_hybatmDA on WCOSS2
~- [ ] C96C48_hybatmDA on Orion~ (not possible at the moment)
- [x] C96C48_ufs_hybatmDA on Hera
- [x] C48mx500_3DVarAOWCDA on Hera
- [x] C192C96 case on ~Hera~Hercules
- [x] C384C192 case on ~Hera~Hercules
- [x] C192C96 case on WCOSS2
- [x] C384C192 case on WCOSS2
- [x] C768C384 case on WCOSS2

# Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes

### CODE REVIEW

**Comment by DavidHuber-NOAA** [comment]
@InnocentSouopgui-NOAA after I have tested this on a couple machines, would you be able to test a forecast-only and cycled case on S4 and Jet?  I'll let you know when it is ready to go.

**Comment by InnocentSouopgui-NOAA** [comment]
> @InnocentSouopgui-NOAA after I have tested this on a couple machines, would you be able to test a forecast-only and cycled case on S4 and Jet? I'll let you know when it is ready to go.

Sure, let me know when it is ready to test on S4.

File: parm/config/gfs/config.anal
Comment (WalterKolczynski-NOAA) [inline parm/config/gfs/config.anal]:
```suggestion
for resource in npe nth; do
  var="${resource}_anal_${RUN}"
  declare -x "${var_type}_anal"="${!var}"
done
declare -x npe_gsi=${npe_anal} # Why does the job use `gsi` and not `anal`?
```

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]

File: parm/config/gfs/config.anal
Comment (DavidHuber-NOAA) [reply parm/config/gfs/config.anal]: Thanks for the suggestion, @WalterKolczynski-NOAA.  The echgres job also reads this config, so I made a change to strip `enkf` from `RUN` so `npe_gsi` can be set.

File: parm/config/gfs/config.resources
Comment (aerorahul) [inline parm/config/gfs/config.resources]: Should we stay with one per line?

Comment (aerorahul) [reply parm/config/gfs/config.resources]: Should these be `unset` at the end of the script?  Seems they are used sometimes.

File: parm/config/gfs/config.resources
Comment (aerorahul) [inline parm/config/gfs/config.resources]: one per line?

File: parm/config/gfs/config.resources
Comment (aerorahul) [inline parm/config/gfs/config.resources:729]: are these here and above intended to remain as `_gfs` or `_gdas`? or should they be `_${RUN}` since they are on the RHS.

File: parm/config/gfs/config.resources.GAEA
Comment (aerorahul) [inline parm/config/gfs/config.resources.GAEA]:
```suggestion
#! /usr/bin/env bash
```
Nothing wrong w/ what you have, but I think we are going with the above.

**Review by aerorahul — COMMENTED** [review]
some comments from a 1st pass.
I will work on the rocoto part where `self._configs` is a matrix of `[RUN][taskname]`.  
Will see if it is ready, but shouldn't hold this PR up.

File: parm/config/gfs/config.resources
Comment (DavidHuber-NOAA) [reply parm/config/gfs/config.resources:729]: Yeah, they need to stay `_gfs` and `_gdas` for now since `config.ufs` only defines variables with `_gfs` and `_gdas` for both deterministic (`fcst`) and ensemble (`efcs`) jobs.  Thus, there isn't a, e.g., `layout_x_enkfgdas` variable to assign to `layout_x`.

File: parm/config/gfs/config.resources
Comment (DavidHuber-NOAA) [reply parm/config/gfs/config.resources]: Yeah, that's probably better practice and more readable.  Fixed.

Comment (DavidHuber-NOAA) [reply parm/config/gfs/config.resources]: Since these aren't `export`'d, they don't need to be.  However, I could see a situation where one of the `config.resources.MACHINE` files would need to set its own value for these, so I'll go ahead and unset them.

Comment (DavidHuber-NOAA) [reply parm/config/gfs/config.resources]: Fixed.

File: parm/config/gfs/config.resources
Comment (aerorahul) [reply parm/config/gfs/config.resources:729]: Thanks!

File: parm/config/gfs/config.resources.JET
Comment (aerorahul) [inline parm/config/gfs/config.resources.JET]: Should this be in `config.resources` (case statement block of line 43) and be called `node_mem_max`?

**Review by aerorahul — APPROVED** [review]
Changes look good.
I left a few comments that could be addressed here or later.

This PR will need to be tested thoroughly before merging as the changes are extensive.

File: parm/config/gfs/config.resources.JET
Comment (DavidHuber-NOAA) [reply parm/config/gfs/config.resources.JET]: I think I like that solution better and to declare it for each system, even though it's not used for other systems (yet).

File: parm/config/gfs/config.resources
Comment (DavidHuber-NOAA) [inline parm/config/gfs/config.resources:111]: @weihuang-jedi Do you know how much memory AWS has per node?

**Comment by DavidHuber-NOAA** [comment]
@InnocentSouopgui-NOAA This PR is now ready to be tested on S4 and Jet.  Would you mind running some cycled tests on both?  I'm not sure if an S2SW test is possible on Jet, but if so, could you also run a forecast-only case there as well?

**Comment by InnocentSouopgui-NOAA** [comment]
> @InnocentSouopgui-NOAA This PR is now ready to be tested on S4 and Jet. Would you mind running some cycled tests on both? I'm not sure if an S2SW test is possible on Jet, but if so, could you also run a forecast-only case there as well?

I will get on it today, and check if S2SW is available on Jet.

File: parm/config/gfs/config.anal
Comment (WalterKolczynski-NOAA) [inline parm/config/gfs/config.anal]: I think we are using only one slash everywhere else, since `$RUN` will only have `enkf` once anyway.
```suggestion
npe_var="npe_anal_${RUN/enkf}"
```

File: parm/config/gfs/config.anal
Comment (WalterKolczynski-NOAA) [inline parm/config/gfs/config.anal:23]: What happened to `nth_anal`?

File: parm/config/gfs/config.resources
Comment (WalterKolczynski-NOAA) [inline parm/config/gfs/config.resources]: I'm not worried about shaving five minutes on the job request when running low-resolution.
```suggestion
    export wtime_analcalc="00:15:00"
```

File: env/HERA.env
Comment (WalterKolczynski-NOAA) [inline env/HERA.env:10]: Thank you for ditching all these.

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]

File: parm/config/gfs/config.anal
Comment (DavidHuber-NOAA) [reply parm/config/gfs/config.anal:23]: Since `nth_anal` and `nth_anal_gfs` were the same, I consolidated them to just `nth_anal` in `config.resources`.

**Comment by InnocentSouopgui-NOAA** [comment]
@DavidHuber-NOAA,
I have a couple of failing tasks on S4, that I am investigating.

`enkfgdasediag` fails silently (job submission fails and the script does not detects), the dependant task `enkfgdaseupd` also fails.
Bellow is the error message from the submission of `enkfgdasediag`, you maybe able to pinpoint the problem quickly.

`gfsfcst` is failing as well, but the error message and where  it fails is different from run to run.

```
+ exglobal_diag.sh[225]: ncmd_max=32
++ exglobal_diag.sh[226]: eval echo srun -l --export=ALL -n '$ncmd' --multi-prog --output=mpmd.%j.%t.out
+++ exglobal_diag.sh[226]: echo srun -l --export=ALL -n 34 --multi-prog --output=mpmd.%j.%t.out
+ exglobal_diag.sh[226]: APRUNCFP_DIAG='srun -l --export=ALL -n 34 --multi-prog --output=mpmd.%j.%t.out'
+ exglobal_diag.sh[227]: srun -l --export=ALL -n 34 --multi-prog --output=mpmd.%j.%t.out /scratch/users/isouopgui/RUNDIRS/test30/ediag.234500/mp_diag.sh
srun: Warning: can't honor --ntasks-per-node set to 32 which doesn't match the requested tasks 34 with the number of requested nodes 2. Ignoring --ntasks-per-node.
srun: Job step aborted: Waiting up to 32 seconds for job step to finish.
slurmstepd: error: Exceeded job memory limit
srun: error: s4-204-c4: tasks 0,5,11-16: Killed
srun: Terminating job step 25889898.0
srun: error: s4-204-c6: tasks 17-28,30-33: Killed
```

**Comment by DavidHuber-NOAA** [comment]
@InnocentSouopgui-NOAA I increased the memory request for the `ediag` job on S4.  See if that works for you.

What resolution(s) are you running the `fcst` at that it fails?  Is there any indication that it may be an out of memory issue?  If not, perhaps there is a node issue?

**Comment by InnocentSouopgui-NOAA** [comment]
> @InnocentSouopgui-NOAA I increased the memory request for the `ediag` job on S4. See if that works for you.
> 
> What resolution(s) are you running the `fcst` at that it fails? Is there any indication that it may be an out of memory issue? If not, perhaps there is a node issue?

`enkfgdasediag` succeds with the update.
`gfsfcst` is still failing.

I can't ping point what is going wrong. Looking at the logfile, the ufs executatble itself seems to run until the end.

The only error message I find towards the end of the logfile is 
```
<gw>/ush/forecast_postdet.sh: line 245: restart_dates[@]: unbound variable
```
I am not having that message on Jet. So it looks like there is setting on S4 or a missing setting causing the problem.
If you want to have a look, the logfile are at `/scratch/users/isouopgui/test/test30/logs/2021110900`

**Comment by InnocentSouopgui-NOAA** [comment]
On S4, resolution 192/96, upp tasks are failing because of openMP over-allocation of resources.
What will be the best place to set `OMP_NUM_THREADS` to prevent that?
I see a couple of possibilities:
1) Explicitly set `OMP_NUM_THREADS` for UPP tasks; For instance, there is a variable `NTHREADS_UPP` computed in <gw>/env/*.env, but I don't find it used anywhere. It looks like the best candidate for `OMP_NUM_THREADS` in UPP tasks. It can be used to set `OMP_NUM_THREADS` right there in <gw>/env/*.env
2) set `OMP_NUM_THREADS` to 1 early, and let tasks that rely on OpenMP set it when they run;
3) A solution for S4 only that uses any of the above.

**Comment by DavidHuber-NOAA** [comment]
@InnocentSouopgui-NOAA Could you try `rocotoboot`ing the job again?  The run directory (`/scratch/users/isouopgui/RUNDIRS/test30/gfsfcst.2021110900/fcst.115872`) no longer exists.  Thanks.

**Comment by DavidHuber-NOAA** [comment]
> On S4, resolution 192/96, upp tasks are failing because of openMP over-allocation of resources. What will be the best place to set `OMP_NUM_THREADS` to prevent that?

@InnocentSouopgui-NOAA That is strange that the UPP is attempting to run threaded.  I suspect this is an issue with the `sbatch` running on S4 not properly picking up the thread count from rocoto.  I think your first solution of adding it to `env/S4.env` is the right way to go as I have not seen this on any other system.

**Comment by InnocentSouopgui-NOAA** [comment]
> @InnocentSouopgui-NOAA Could you try `rocotoboot`ing the job again? The run directory (`/scratch/users/isouopgui/RUNDIRS/test30/gfsfcst.2021110900/fcst.115872`) no longer exists. Thanks.

Done.

**Comment by DavidHuber-NOAA** [comment]
> > @InnocentSouopgui-NOAA Could you try `rocotoboot`ing the job again? The run directory (`/scratch/users/isouopgui/RUNDIRS/test30/gfsfcst.2021110900/fcst.115872`) no longer exists. Thanks.
> 
> Done.

@InnocentSouopgui-NOAA It looks like the job failed because the inputs for the gfsfcst are no longer available:
```
 FATAL ERROR: Cold start ICs are missing from '/scratch/users/isouopgui/test/test30/gfs.20211109/00//model_data/atmos/input'
```

Perhaps you could try rebooting the last gfsfcst (for cycle 202111110000)?

**Comment by InnocentSouopgui-NOAA** [comment]
> > > @InnocentSouopgui-NOAA Could you try `rocotoboot`ing the job again? The run directory (`/scratch/users/isouopgui/RUNDIRS/test30/gfsfcst.2021110900/fcst.115872`) no longer exists. Thanks.
> > 
> > 
> > Done.
> 
> @InnocentSouopgui-NOAA It looks like the job failed because the inputs for the gfsfcst are no longer available:
> 
> ```
>  FATAL ERROR: Cold start ICs are missing from '/scratch/users/isouopgui/test/test30/gfs.20211109/00//model_data/atmos/input'
> ```
> 
> Perhaps you could try rebooting the last gfsfcst (for cycle 202111110000)?

It's a little trickier. What you saw is the error message from the third attempt, which is different from the error message from the first attempt. Have a look at the error message from the first attempt.
gfsfcst is failing in all run so far on S4. However, the failure does not seem to affect subsequent tasks and cycles.


The only error message that I find in the logfile from the first attempt is the following:
```
/ush/forecast_postdet.sh: line 245: restart_dates[@]: unbound variable
```
It looks like S4 is not happy with the restart_dates[@] when it is empty. That is my suspicion for now.

**Comment by DavidHuber-NOAA** [comment]
@InnocentSouopgui-NOAA I see now, thanks for explaining that again for me.  So, this looks like an issue with bash interpreters.  The ones on Hera, Hercules, etc are all newer than S4 and treat references to empty arrays as valid references, but S4's version treats them as unbound variables.  The way around this is to add a check for an empty array before running the for loop.

Problem code
https://github.com/NOAA-EMC/global-workflow/blob/8993b42cb91144c0ab0501dc7841ea8d675c4701/ush/forecast_postdet.sh#L257-L284

Fix
```bash
  restart_dates=()
  # Copy restarts in the assimilation window for RUN=gdas|enkfgdas|enkfgfs
  if [[ "${RUN}" =~ "gdas" || "${RUN}" == "enkfgfs" ]]; then
    restart_date="${model_start_date_next_cycle}"
    while (( restart_date <= forecast_end_cycle )); do
      restart_dates+=("${restart_date:0:8}.${restart_date:8:2}0000")
      restart_date=$(date --utc -d "${restart_date:0:8} ${restart_date:8:2} + ${restart_interval} hours" +%Y%m%d%H)
    done
  elif [[ "${RUN}" == "gfs" || "${RUN}" == "gefs" ]]; then # Copy restarts at the end of the forecast segment for RUN=gfs|gefs
    if [[ "${COPY_FINAL_RESTARTS}" == "YES" ]]; then
      restart_dates+=("${forecast_end_cycle:0:8}.${forecast_end_cycle:8:2}0000")
    fi
  fi

  ### Check that there are restart files to copy
  if [[ ${#restart_dates} -gt 0 ]]; then
    # Get list of FV3 restart files
    local file_list fv3_file
    file_list=$(FV3_restarts)

    # Copy restarts for the dates collected above to COM
    for restart_date in "${restart_dates[@]}"; do
      echo "Copying FV3 restarts for 'RUN=${RUN}' at ${restart_date}"
      for fv3_file in ${file_list}; do
        ${NCP} "${DATArestart}/FV3_RESTART/${restart_date}.${fv3_file}" \
               "${COMOUT_ATMOS_RESTART}/${restart_date}.${fv3_file}"
      done
    done

    echo "SUB ${FUNCNAME[0]}: Output data for FV3 copied"
  fi
}
```

**Comment by DavidHuber-NOAA** [comment]
@InnocentSouopgui-NOAA I have put the fix in place.  Can you check it again on S4?  You'll probably need to restart the experiment completely.

**Comment by InnocentSouopgui-NOAA** [comment]
I successfully ran a couple of tests on S4 and Jet.
- On S4, cycle at resolution 96/48 and 192/96
- On Jet:
   - Cycle at resolution 192/96
   - forecast only at resolution 96 and 384

I submitted a pull request on @DavidHuber-NOAA global workflow repos with a couple of modifications I needed to get thinks running smoothly.

**Review by WalterKolczynski-NOAA — APPROVED** [review]

**Comment by DavidHuber-NOAA** [comment]
Fixed an issue with the atmensanlfv3inc declaration in config.resources that was causing the job to fail on WCOSS2.  Relaunching WCOSS2 CI.

**Comment by WalterKolczynski-NOAA** [comment]
None of the gempak jobs ran on WCOSS even thought the dependencies are satisfied:
```
WCOSS2 (BACKUPSYS) C96_atm3DVar_extended_184d2359> rocotocheck -d C96_atm3DVar_extended_184d2359.db -w C96_atm3DVar_extended_184d2359.xml -t gdasgempak -c 202112210000

Task: gdasgempak
  account: GFS-DEV
  command: /lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2672/global-workflow/jobs/rocoto/gempak.sh
  cores: 2
  cycledefs: gdas
  final: false
  jobname: C96_atm3DVar_extended_184d2359_gdasgempak_00
  join: /lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2672/RUNTESTS/COMROOT/C96_atm3DVar_extended_184d2359/logs/2021122100/gdasgempak.log
  maxtries: 2
  memory: 4GB
  name: gdasgempak
  nodes: 1:ppn=2:tpp=
  queue: dev
  throttle: 9999999
  walltime: 03:00:00
  environment
    CDATE ==> 2021122100
    CDUMP ==> gdas
    COMROOT ==> /lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2672/RUNTESTS/COMROOT
    DATAROOT ==> /lfs/h2/emc/stmp/terry.mcguinness/RUNDIRS/C96_atm3DVar_extended_184d2359
    EXPDIR ==> /lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2672/RUNTESTS/EXPDIR/C96_atm3DVar_extended_184d2359
    HOMEgfs ==> /lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2672/global-workflow
    NET ==> gfs
    PDY ==> 20211221
    RUN ==> gdas
    RUN_ENVIR ==> emc
    cyc ==> 00
  dependencies
    SOME is satisfied
      gdasatmos_prod_f000 of cycle 202112210000 is SUCCEEDED
      gdasatmos_prod_f003 of cycle 202112210000 is SUCCEEDED
      gdasatmos_prod_f006 of cycle 202112210000 is SUCCEEDED
      gdasatmos_prod_f009 of cycle 202112210000 is SUCCEEDED

Cycle: 202112210000
  Valid for this task: YES
  State: active
  Activated: 2024-06-23 11:55:17 UTC
  Completed: -
  Expired: -

Job: This task has not been submitted for this cycle

Task can not be submitted because:
```

Is the job specification malformed, preventing submission to torque?

**Comment by DavidHuber-NOAA** [comment]
@WalterKolczynski-NOAA Looking at the rocoto log file `/lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2672/RUNTESTS/EXPDIR/C96_atm3DVar_extended_184d2359/logs/2021122100.log`, I am seeing the message
```
2220 2024-06-23 19:50:13 +0000 :: dlogin09 :: Submission of gdasgempak failed!  qsub: Unknown resource: tpp
2221 Job submit error: 187.
```

Looking in the xml, I see
```xml
<task name="gdasgempak" cycledefs="gdas" maxtries="&MAXTRIES;">

   <command>/lfs/h2/emc/global/noscrub/globalworkflow.ci/GFS_CI_ROOT/PR/2672/global-workflow/jobs/rocoto/gempak.sh</command>

   <jobname><cyclestr>C96_atm3DVar_extended_184d2359_gdasgempak_@H</cyclestr></jobname>
   <account>GFS-DEV</account>
   <queue>dev</queue>
   <walltime>03:00:00</walltime>
   <nodes>1:ppn=2:tpp=</nodes>
```

So it seems `tpp` is not being set.  Looking in config.resources, I see the issue:

```bash
  "gempak")
    export wtime_gempak="03:00:00"
    export npe_gempak_gdas=2
    export npe_gempak_gfs=28
    export npe_node_gempak_gdas=2
    export npe_node_gempak_gfs=28
    export nth_gempak=1
    export memory_gempak_gdas="4GB"
    export memory_gempak_gfs="2GB"

    var_npe_node="npe_node_gempak_${RUN}"
    var_nth="nth_gempak_${RUN}"
    var_npe="npe_gempak_${RUN}"
    # RUN is set to a single value at setup time, so these won't be found
    # TODO rework setup_xml.py to initialize RUN to the applicable option
    if [[ -n "${!var_npe_node+0}" ]]; then
      declare -x "npe_node_gempak"="${!var_npe_node}" \
                 "nth_gempak"="${!var_nth}" \                  
                 "npe_gempak"="${!var_npe}"
    fi
    ;;
```

So `nth_gempak` is being set to `${nth_gempak_gdas}`, which is undefined.  I removed this assignment and am now rerunning the `C96_atm3DVar_extended` test on WCOSS2.

**Comment by DavidHuber-NOAA** [comment]
`gdasgempak` and `gdasgempakmetancdc` completed successfully on the first full cycle:
```
       CYCLE                    TASK                       JOBID               STATE         EXIT STATUS     TRIES      DURATION
================================================================================================================================
202112210000              gdasgempak                   153723934           SUCCEEDED                   0         1         329.0
202112210000      gdasgempakmetancdc                   153724367           SUCCEEDED                   0         1          27.0

**Comment by DavidHuber-NOAA** [comment]
@WalterKolczynski-NOAA The manual `C96_atm3DVar_extended` test passed on WCOSS2.

**Pull request merged**
